File: plugins/custom-native/ios/Sources/NativeKitCustomPlugin/NativeKitCustomPlugin.swift
import Foundation
import Capacitor
import CoreLocation
import UIKit
import WebKit

@objc(NativeKitCustomPlugin)
public class NativeKitCustomPlugin: CAPPlugin, CAPBridgedPlugin, CLLocationManagerDelegate {
    public let identifier = "NativeKitCustomPlugin"
    public let jsName = "NativeKitCustom"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkAlarmCapabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestExactAlarmAccess", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestFullScreenIntentAccess", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "scheduleAlarm", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelAlarm", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listAlarms", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopRinging", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startSSE", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopSSE", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secureSet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secureGet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secureRemove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secureClear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startBackgroundLocation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopBackgroundLocation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBackgroundLocationStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBufferedLocations", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearBufferedLocations", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAppSettings", returnType: CAPPluginReturnPromise),
    ]

    private final class StreamHandle {
        let session: URLSession
        let task: URLSessionDataTask
        let delegate: NativeStreamDelegate
        init(session: URLSession, task: URLSessionDataTask, delegate: NativeStreamDelegate) {
            self.session = session; self.task = task; self.delegate = delegate
        }
    }

    private let secureStore = NativeKitSecureStore()
    private let alarmAdapter = NativeKitAlarmAdapter()
    private var streams: [String: StreamHandle] = [:]
    private var locationManager: CLLocationManager?
    private var maxLocationBuffer = 100
    private let locationDefaultsKey = "nativekit.buffered.locations"

    public override func shouldOverrideLoad(_ navigationAction: WKNavigationAction) -> NSNumber? {
        guard let url = navigationAction.request.url, let bridge else { return nil }
        let config = bridge.config
        let sameOrigin = { (candidate: URL, base: URL) in
            candidate.scheme?.lowercased() == base.scheme?.lowercased()
                && candidate.host?.lowercased() == base.host?.lowercased()
                && candidate.port == base.port
        }
        let localAppURL = sameOrigin(url, config.serverURL) || sameOrigin(url, config.localURL)
        if localAppURL || url.absoluteString.lowercased() == "about:srcdoc" { return nil }
        let scheme = url.scheme?.lowercased()
        if (scheme == "http" || scheme == "https"),
           let host = url.host,
           config.shouldAllowNavigation(to: host) { return nil }
        // Prevent opaque App Browser frames from resetting their CSP through data/blob navigation,
        // making an unbrokered network request, or launching an external application. Trusted code
        // should use an explicit native API (for example the Browser plugin) for external navigation.
        return NSNumber(value: true)
    }

    @objc func checkAlarmCapabilities(_ call: CAPPluginCall) {
        call.resolve(alarmAdapter.capabilities())
    }

    @objc func requestExactAlarmAccess(_ call: CAPPluginCall) {
        alarmAdapter.requestAuthorization { result in
            DispatchQueue.main.async {
                switch result { case .success: call.resolve(); case .failure(let error): call.reject(error.localizedDescription, nil, error) }
            }
        }
    }

    @objc func requestFullScreenIntentAccess(_ call: CAPPluginCall) {
        call.resolve() // iOS does not expose Android-style full-screen intent access.
    }

    @objc func scheduleAlarm(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let title = call.getString("title"), let at = call.getDouble("at") else {
            call.reject("id, title and numeric at are required"); return
        }
        var options: [String: Any] = [
            "id": id, "title": title, "at": at,
            "body": call.getString("body") ?? "",
            "repeatIntervalMinutes": call.getDouble("repeatIntervalMinutes") ?? 0,
            "fullScreen": call.getBool("fullScreen") ?? false,
            "sound": call.getString("sound") ?? "default",
        ]
        if let extra = call.getObject("extra") { options["extra"] = extra }
        alarmAdapter.schedule(options: options) { result in
            DispatchQueue.main.async {
                switch result { case .success(let value): call.resolve(value); case .failure(let error): call.reject(error.localizedDescription, nil, error) }
            }
        }
    }

    @objc func cancelAlarm(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else { call.reject("id is required"); return }
        do { try alarmAdapter.cancel(id: id); call.resolve() }
        catch { call.reject(error.localizedDescription, nil, error) }
    }

    @objc func listAlarms(_ call: CAPPluginCall) { call.resolve(["alarms": alarmAdapter.list()]) }
    @objc func stopRinging(_ call: CAPPluginCall) { alarmAdapter.stop(id: call.getString("id")); call.resolve() }

    @objc func startSSE(_ call: CAPPluginCall) {
        guard let raw = call.getString("url"), let url = URL(string: raw), ["http", "https"].contains(url.scheme?.lowercased() ?? "") else {
            call.reject("A valid HTTP(S) url is required"); return
        }
        let id = call.getString("streamId") ?? UUID().uuidString
        let format = call.getString("format") ?? "sse"
        var request = URLRequest(url: url)
        request.httpMethod = call.getString("method") ?? "GET"
        request.timeoutInterval = 30
        if let headers = call.getObject("headers") {
            for (key, value) in headers { request.setValue(String(describing: value), forHTTPHeaderField: key) }
        }
        if format == "sse" && request.value(forHTTPHeaderField: "Accept") == nil { request.setValue("text/event-stream", forHTTPHeaderField: "Accept") }
        if let body = call.getString("body") { request.httpBody = body.data(using: .utf8) }

        let delegate = NativeStreamDelegate(streamId: id, format: format, allowRedirects: !(call.getBool("disableRedirects") ?? false))
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 7 * 24 * 60 * 60
        let session = URLSession(configuration: configuration, delegate: delegate, delegateQueue: nil)
        let task = session.dataTask(with: request)
        delegate.onData = { [weak self] value in
            var event = value; event["streamId"] = id
            DispatchQueue.main.async { self?.notifyListeners("nativeSSEData", data: event, retainUntilConsumed: true) }
        }
        delegate.onEnd = { [weak self] status in
            DispatchQueue.main.async {
                guard let self, self.streams.removeValue(forKey: id) != nil else { return }
                self.notifyListeners("nativeSSEEnd", data: ["streamId": id, "status": status as Any], retainUntilConsumed: true)
                session.finishTasksAndInvalidate()
            }
        }
        delegate.onError = { [weak self] message, status in
            DispatchQueue.main.async {
                guard let self, self.streams.removeValue(forKey: id) != nil else { return }
                self.notifyListeners("nativeSSEError", data: ["streamId": id, "message": message, "status": status as Any], retainUntilConsumed: true)
                session.invalidateAndCancel()
            }
        }
        streams[id] = StreamHandle(session: session, task: task, delegate: delegate)
        task.resume()
        call.resolve(["streamId": id])
    }

    @objc func stopSSE(_ call: CAPPluginCall) {
        guard let id = call.getString("streamId") else { call.reject("streamId is required"); return }
        if let handle = streams.removeValue(forKey: id) { handle.task.cancel(); handle.session.invalidateAndCancel() }
        call.resolve()
    }

    @objc func secureSet(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let value = call.getString("value") else { call.reject("key and value are required"); return }
        do { try secureStore.set(key: key, value: value); call.resolve() } catch { call.reject("Secure storage write failed", nil, error) }
    }
    @objc func secureGet(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else { call.reject("key is required"); return }
        do { call.resolve(["value": try secureStore.get(key: key) as Any]) } catch { call.reject("Secure storage read failed", nil, error) }
    }
    @objc func secureRemove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else { call.reject("key is required"); return }
        do { try secureStore.remove(key: key); call.resolve() } catch { call.reject("Secure storage delete failed", nil, error) }
    }
    @objc func secureClear(_ call: CAPPluginCall) {
        do { try secureStore.clear(); call.resolve() } catch { call.reject("Secure storage clear failed", nil, error) }
    }

    @objc func startBackgroundLocation(_ call: CAPPluginCall) {
        let manager = locationManager ?? CLLocationManager()
        locationManager = manager
        manager.delegate = self
        let status = manager.authorizationStatus
        if status == .notDetermined { manager.requestAlwaysAuthorization(); call.reject("Location authorization requested; approve it and retry"); return }
        if status == .authorizedWhenInUse { manager.requestAlwaysAuthorization(); call.reject("Choose Always Allow, then retry background location"); return }
        guard status == .authorizedAlways else { call.reject("Always location permission is required"); return }
        manager.desiredAccuracy = (call.getString("desiredAccuracy") ?? "high") == "high" ? kCLLocationAccuracyBest : kCLLocationAccuracyHundredMeters
        manager.distanceFilter = call.getDouble("minDistanceM") ?? 10
        manager.pausesLocationUpdatesAutomatically = true
        manager.allowsBackgroundLocationUpdates = true
        manager.showsBackgroundLocationIndicator = true
        maxLocationBuffer = max(10, min(1000, call.getInt("maxBuffer") ?? 100))
        manager.startUpdatingLocation()
        UserDefaults.standard.set(true, forKey: "nativekit.location.running")
        call.resolve(["running": true])
    }

    @objc func stopBackgroundLocation(_ call: CAPPluginCall) {
        locationManager?.stopUpdatingLocation()
        locationManager?.allowsBackgroundLocationUpdates = false
        UserDefaults.standard.set(false, forKey: "nativekit.location.running")
        call.resolve()
    }

    @objc func getBackgroundLocationStatus(_ call: CAPPluginCall) {
        let status = locationManager?.authorizationStatus ?? CLLocationManager().authorizationStatus
        let permission: String
        switch status { case .authorizedAlways: permission = "granted"; case .authorizedWhenInUse: permission = "foreground-only"; case .notDetermined: permission = "prompt"; default: permission = "denied" }
        call.resolve(["running": UserDefaults.standard.bool(forKey: "nativekit.location.running"), "permission": permission, "platform": "ios"])
    }

    @objc func getBufferedLocations(_ call: CAPPluginCall) {
        call.resolve(["locations": UserDefaults.standard.array(forKey: locationDefaultsKey) ?? []])
    }

    @objc func clearBufferedLocations(_ call: CAPPluginCall) {
        UserDefaults.standard.removeObject(forKey: locationDefaultsKey); call.resolve()
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        for location in locations {
            var value: [String: Any] = [
                "latitude": location.coordinate.latitude,
                "longitude": location.coordinate.longitude,
                "accuracy": location.horizontalAccuracy,
                "timestamp": location.timestamp.timeIntervalSince1970 * 1000,
            ]
            if location.verticalAccuracy >= 0 { value["altitude"] = location.altitude }
            if location.speed >= 0 { value["speed"] = location.speed }
            if location.course >= 0 { value["bearing"] = location.course }
            var buffered = UserDefaults.standard.array(forKey: locationDefaultsKey) as? [[String: Any]] ?? []
            buffered.append(value)
            if buffered.count > maxLocationBuffer { buffered.removeFirst(buffered.count - maxLocationBuffer) }
            UserDefaults.standard.set(buffered, forKey: locationDefaultsKey)
            notifyListeners("nativeLocation", data: value, retainUntilConsumed: true)
        }
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        notifyListeners("nativeLocationError", data: ["message": error.localizedDescription], retainUntilConsumed: true)
    }

    @objc func openAppSettings(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString), UIApplication.shared.canOpenURL(url) else { call.reject("Unable to open app settings"); return }
            UIApplication.shared.open(url) { opened in opened ? call.resolve() : call.reject("Unable to open app settings") }
        }
    }

    deinit {
        for (_, handle) in streams { handle.session.invalidateAndCancel() }
        locationManager?.stopUpdatingLocation()
    }
}

File: plugins/custom-native/android/src/main/java/dev/nativekit/custom/NativeKitCustomPlugin.java
package dev.nativekit.custom;

import android.Manifest;
import android.app.AlarmManager;
import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import androidx.core.content.ContextCompat;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Headers;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.IOException;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

@CapacitorPlugin(name = "NativeKitCustom")
public class NativeKitCustomPlugin extends Plugin {
    private final Map<String, Call> streams = new ConcurrentHashMap<>();
    private OkHttpClient http;
    private SecureStore secureStore;
    private BroadcastReceiver eventReceiver;

    @Override public void load() {
        secureStore = new SecureStore(getContext());
        http = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .writeTimeout(120, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .build();
        eventReceiver = new BroadcastReceiver() {
            @Override public void onReceive(Context context, Intent intent) {
                if (AlarmReceiver.FIRED_BROADCAST.equals(intent.getAction())) {
                    JSObject value = new JSObject();
                    value.put("id", intent.getStringExtra("alarmId"));
                    notifyListeners("nativeAlarmFired", value, true);
                } else if (LocationTrackingService.LOCATION_BROADCAST.equals(intent.getAction())) {
                    try { notifyListeners("nativeLocation", new JSObject(intent.getStringExtra("location")), true); }
                    catch (Exception ignored) {}
                }
            }
        };
        IntentFilter filter = new IntentFilter();
        filter.addAction(AlarmReceiver.FIRED_BROADCAST);
        filter.addAction(LocationTrackingService.LOCATION_BROADCAST);
        if (Build.VERSION.SDK_INT >= 33) getContext().registerReceiver(eventReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
        else getContext().registerReceiver(eventReceiver, filter);
    }

    @Override public Boolean shouldOverrideLoad(Uri url) {
        if (url == null) return null;
        String scheme = url.getScheme();
        Uri appUri = Uri.parse(getBridge().getAppUrl());
        boolean localAppUrl = appUri.getHost() != null && appUri.getHost().equalsIgnoreCase(url.getHost())
            && appUri.getScheme() != null && appUri.getScheme().equalsIgnoreCase(scheme)
            && appUri.getPort() == url.getPort();
        if (localAppUrl) return null;
        if (("http".equalsIgnoreCase(scheme) || "https".equalsIgnoreCase(scheme))
            && url.getHost() != null && getBridge().getAppAllowNavigationMask().matches(url.getHost())) return null;
        if ("about:srcdoc".equalsIgnoreCase(url.toString())) return null;
        // Prevent opaque App Browser frames from resetting their CSP through data/blob navigation,
        // making an unbrokered network request, or launching an external intent. Trusted code should
        // use an explicit native API (for example the Browser plugin) for external navigation.
        return true;
    }

    @Override protected void handleOnDestroy() {
        for (Call call : streams.values()) call.cancel();
        streams.clear();
        if (eventReceiver != null) {
            try { getContext().unregisterReceiver(eventReceiver); } catch (Exception ignored) {}
        }
        super.handleOnDestroy();
    }

    @PluginMethod public void checkAlarmCapabilities(PluginCall call) {
        AlarmScheduler scheduler = new AlarmScheduler(getContext());
        boolean fullScreen = true;
        if (Build.VERSION.SDK_INT >= 34) fullScreen = getContext().getSystemService(NotificationManager.class).canUseFullScreenIntent();
        JSObject result = new JSObject();
        result.put("platform", "android");
        result.put("exact", scheduler.canScheduleExact());
        result.put("fullScreen", fullScreen);
        result.put("alarmKit", false);
        result.put("fallback", scheduler.canScheduleExact() ? "none" : "inexact-native-alarm");
        call.resolve(result);
    }

    @PluginMethod public void requestExactAlarmAccess(PluginCall call) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S || new AlarmScheduler(getContext()).canScheduleExact()) { call.resolve(); return; }
        try {
            Intent intent = new Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM, Uri.parse("package:" + getContext().getPackageName()));
            getActivity().startActivity(intent);
            call.resolve();
        } catch (Exception error) { call.reject("Unable to open exact alarm settings", error); }
    }

    @PluginMethod public void requestFullScreenIntentAccess(PluginCall call) {
        if (Build.VERSION.SDK_INT < 34) { call.resolve(); return; }
        try {
            Intent intent = new Intent(Settings.ACTION_MANAGE_APP_USE_FULL_SCREEN_INTENT, Uri.parse("package:" + getContext().getPackageName()));
            getActivity().startActivity(intent);
            call.resolve();
        } catch (Exception error) { call.reject("Unable to open full-screen intent settings", error); }
    }

    @PluginMethod public void scheduleAlarm(PluginCall call) {
        String id = required(call, "id");
        String title = required(call, "title");
        Long at = call.getLong("at");
        if (id == null || title == null || at == null) { call.reject("id, title and numeric at are required"); return; }
        try {
            JSONObject alarm = new JSONObject();
            alarm.put("id", id);
            alarm.put("title", title);
            alarm.put("body", call.getString("body", ""));
            alarm.put("scheduledAt", at);
            alarm.put("repeatIntervalMinutes", call.getLong("repeatIntervalMinutes", 0L));
            alarm.put("fullScreen", call.getBoolean("fullScreen", false));
            alarm.put("sound", call.getString("sound", "default"));
            JSObject extra = call.getObject("extra");
            if (extra != null) alarm.put("extra", extra);
            JSONObject result = new AlarmScheduler(getContext()).schedule(alarm, true);
            call.resolve(new JSObject(result.toString()));
        } catch (Exception error) { call.reject("Alarm could not be scheduled: " + error.getMessage(), error); }
    }

    @PluginMethod public void cancelAlarm(PluginCall call) {
        String id = required(call, "id");
        if (id == null) return;
        new AlarmScheduler(getContext()).cancel(id);
        call.resolve();
    }

    @PluginMethod public void listAlarms(PluginCall call) {
        try {
            JSObject result = new JSObject();
            result.put("alarms", new JSArray(new AlarmScheduler(getContext()).listJson().toString()));
            call.resolve(result);
        } catch (Exception error) {
            call.reject("Stored alarms could not be read", error);
        }
    }

    @PluginMethod public void stopRinging(PluginCall call) {
        getContext().startService(new Intent(getContext(), AlarmService.class).setAction(AlarmService.ACTION_STOP).putExtra("alarmId", call.getString("id", "")));
        call.resolve();
    }

    @PluginMethod public void startSSE(PluginCall pluginCall) {
        String url = required(pluginCall, "url");
        if (url == null) return;
        Uri parsed = Uri.parse(url);
        if (!("https".equals(parsed.getScheme()) || "http".equals(parsed.getScheme()))) { pluginCall.reject("Only HTTP(S) streams are allowed"); return; }
        String streamId = pluginCall.getString("streamId", UUID.randomUUID().toString());
        String method = pluginCall.getString("method", "GET").toUpperCase();
        String format = pluginCall.getString("format", "sse");
        Request.Builder builder = new Request.Builder().url(url);
        JSObject headers = pluginCall.getObject("headers");
        if (headers != null) {
            java.util.Iterator<String> keys = headers.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                builder.addHeader(key, String.valueOf(headers.opt(key)));
            }
        }
        if ("sse".equals(format) && (headers == null || !headers.has("Accept"))) builder.addHeader("Accept", "text/event-stream");
        String body = pluginCall.getString("body");
        if (!"GET".equals(method) && !"HEAD".equals(method)) {
            MediaType type = MediaType.parse(headers == null ? "application/json; charset=utf-8" : headers.optString("Content-Type", "application/json; charset=utf-8"));
            builder.method(method, RequestBody.create(body == null ? "" : body, type));
        } else builder.method(method, null);

        OkHttpClient streamClient = pluginCall.getBoolean("disableRedirects", false)
            ? http.newBuilder().followRedirects(false).followSslRedirects(false).build()
            : http;
        Call stream = streamClient.newCall(builder.build());
        streams.put(streamId, stream);
        stream.enqueue(new Callback() {
            @Override public void onFailure(Call call, IOException error) {
                if (streams.remove(streamId) == null) return;
                JSObject event = streamEvent(streamId, format);
                event.put("message", error.getMessage());
                notifyListeners("nativeSSEError", event, true);
            }

            @Override public void onResponse(Call call, Response response) {
                try (ResponseBody responseBody = response.body()) {
                    if (!response.isSuccessful() || responseBody == null) {
                        JSObject event = streamEvent(streamId, format);
                        event.put("message", "HTTP " + response.code());
                        event.put("status", response.code());
                        notifyListeners("nativeSSEError", event, true);
                        return;
                    }
                    if ("text".equals(format)) readText(streamId, responseBody.source(), format);
                    else readLines(streamId, responseBody.source(), format);
                    if (streams.containsKey(streamId)) {
                        JSObject event = streamEvent(streamId, format);
                        event.put("status", response.code());
                        notifyListeners("nativeSSEEnd", event, true);
                    }
                } catch (Exception error) {
                    if (streams.containsKey(streamId)) {
                        JSObject event = streamEvent(streamId, format);
                        event.put("message", String.valueOf(error.getMessage()));
                        event.put("status", response.code());
                        notifyListeners("nativeSSEError", event, true);
                    }
                } finally { streams.remove(streamId); }
            }
        });
        JSObject result = new JSObject();
        result.put("streamId", streamId);
        pluginCall.resolve(result);
    }

    private void readText(String id, BufferedSource source, String format) throws IOException {
        Buffer buffer = new Buffer();
        while (streams.containsKey(id) && source.read(buffer, 8192) != -1) {
            String chunk = buffer.readUtf8();
            JSObject event = streamEvent(id, format);
            event.put("data", chunk);
            notifyListeners("nativeSSEData", event, true);
        }
    }

    private void readLines(String id, BufferedSource source, String format) throws IOException {
        if ("ndjson".equals(format)) {
            String line;
            while (streams.containsKey(id) && (line = source.readUtf8Line()) != null) {
                if (!line.trim().isEmpty()) {
                    JSObject event = streamEvent(id, format); event.put("data", line); notifyListeners("nativeSSEData", event, true);
                }
            }
            return;
        }
        StringBuilder data = new StringBuilder();
        String eventName = null;
        String eventId = null;
        String line;
        while (streams.containsKey(id) && (line = source.readUtf8Line()) != null) {
            if (line.isEmpty()) {
                if (data.length() > 0 || eventName != null || eventId != null) {
                    JSObject event = streamEvent(id, format);
                    event.put("data", data.length() > 0 ? data.substring(0, data.length() - 1) : "");
                    if (eventName != null) event.put("event", eventName);
                    if (eventId != null) event.put("id", eventId);
                    notifyListeners("nativeSSEData", event, true);
                }
                data.setLength(0); eventName = null; eventId = null; continue;
            }
            if (line.startsWith(":")) continue;
            int colon = line.indexOf(':');
            String field = colon < 0 ? line : line.substring(0, colon);
            String value = colon < 0 ? "" : line.substring(colon + 1).replaceFirst("^ ", "");
            if ("data".equals(field)) data.append(value).append('\n');
            else if ("event".equals(field)) eventName = value;
            else if ("id".equals(field)) eventId = value;
        }
    }

    @PluginMethod public void stopSSE(PluginCall call) {
        String id = required(call, "streamId");
        if (id == null) return;
        Call stream = streams.remove(id);
        if (stream != null) stream.cancel();
        call.resolve();
    }

    @PluginMethod public void secureSet(PluginCall call) {
        String key = required(call, "key"); String value = required(call, "value");
        if (key == null || value == null) return;
        try { secureStore.set(key, value); call.resolve(); } catch (Exception error) { call.reject("Secure storage write failed", error); }
    }

    @PluginMethod public void secureGet(PluginCall call) {
        String key = required(call, "key"); if (key == null) return;
        try { JSObject result = new JSObject(); result.put("value", secureStore.get(key)); call.resolve(result); }
        catch (Exception error) { call.reject("Secure storage read failed", error); }
    }

    @PluginMethod public void secureRemove(PluginCall call) { String key = required(call, "key"); if (key != null) { secureStore.remove(key); call.resolve(); } }
    @PluginMethod public void secureClear(PluginCall call) { secureStore.clear(); call.resolve(); }

    @PluginMethod public void startBackgroundLocation(PluginCall call) {
        boolean foreground = ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;
        if (!foreground) { call.reject("Foreground location permission is required first"); return; }
        if (Build.VERSION.SDK_INT >= 29 && ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            call.reject("Grant 'Allow all the time' location access in app settings before starting background tracking"); return;
        }
        Intent service = new Intent(getContext(), LocationTrackingService.class).setAction(LocationTrackingService.ACTION_START)
            .putExtra("minTimeMs", call.getLong("minTimeMs", 30_000L))
            .putExtra("minDistanceM", call.getFloat("minDistanceM", 10f))
            .putExtra("maxBuffer", call.getInt("maxBuffer", 100));
        ContextCompat.startForegroundService(getContext(), service);
        JSObject result = new JSObject(); result.put("running", true); call.resolve(result);
    }

    @PluginMethod public void stopBackgroundLocation(PluginCall call) {
        getContext().startService(new Intent(getContext(), LocationTrackingService.class).setAction(LocationTrackingService.ACTION_STOP));
        call.resolve();
    }

    @PluginMethod public void getBackgroundLocationStatus(PluginCall call) {
        boolean foreground = ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;
        boolean background = Build.VERSION.SDK_INT < 29 || ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_BACKGROUND_LOCATION) == PackageManager.PERMISSION_GRANTED;
        JSObject result = new JSObject();
        result.put("running", getContext().getSharedPreferences("nativekit_location_store", Context.MODE_PRIVATE).getBoolean("running", false));
        result.put("permission", foreground && background ? "granted" : foreground ? "foreground-only" : "denied");
        result.put("platform", "android");
        call.resolve(result);
    }

    @PluginMethod public void getBufferedLocations(PluginCall call) {
        String raw = getContext().getSharedPreferences("nativekit_location_store", Context.MODE_PRIVATE).getString("locations", "[]");
        JSObject result = new JSObject();
        try { result.put("locations", new JSArray(raw)); } catch (Exception ignored) { result.put("locations", new JSArray()); }
        call.resolve(result);
    }

    @PluginMethod public void clearBufferedLocations(PluginCall call) {
        getContext().getSharedPreferences("nativekit_location_store", Context.MODE_PRIVATE).edit().remove("locations").apply();
        call.resolve();
    }

    @PluginMethod public void openAppSettings(PluginCall call) {
        try {
            Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse("package:" + getContext().getPackageName()));
            getActivity().startActivity(intent);
            call.resolve();
        } catch (Exception error) { call.reject("Unable to open app settings", error); }
    }

    private String required(PluginCall call, String key) {
        String value = call.getString(key);
        if (value == null || value.isEmpty()) { call.reject(key + " is required"); return null; }
        return value;
    }

    private JSObject streamEvent(String streamId, String format) {
        JSObject event = new JSObject(); event.put("streamId", streamId); event.put("format", format); return event;
    }
}
